### PR TITLE
Invoice archival

### PR DESCRIPTION
**Pull request by tmb**
Closes #121

### CODE REVIEW

File: app/views/invoices/show.html.erb [hunk 1]
```diff
@@ -92,9 +92,10 @@
     </table>
   </section>
 
-  <section class="card__banner border-top">
+  <section class="card__banner border-top flex">
     <%= link_to 'View sent invoice', @invoice.hosted_invoice_url, class: 'btn bg-accent mr2', target: '_blank'  %>
     <%= link_to 'Download PDF', @invoice.invoice_pdf, class: 'btn bg-info', target: '_blank'  %>
+    <%= link_to "#{!@invoice.archived? ? 'Archive Invoice' : 'Un-archive Invoice'}", (!@invoice.archived? ? invoice_archive_path(@invoice) : invoice_unarchive_path(@invoice)), method: :post, class: 'btn bg-slate ml-auto' %>
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb]: 1. You can remove the outer quotes & interpolation
2. Can remove “Invoice” from the button name
3. Basically never use a bang in a ternary—remove it & reverse the options

File: app/views/invoices/show.html.erb [hunk 2]
```diff
@@ -92,9 +92,10 @@
     </table>
   </section>
 
-  <section class="card__banner border-top">
+  <section class="card__banner border-top flex">
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb:105]: Did you check this on mobile? May need .flex-wrap/other fixes to prevent weirdness.

**Review by lachlanjc — COMMENTED** [review]
Just those two little things. Great job with this PR—markedly better code across frontend + backend than your early contributions, I love seeing this progress. You’re awesome Theo. 💟

**Comment by tmb** [comment]
~just found out that Stripe doesn't allow you to update the due date on invoices :(~ I was looking at this PR thinking it was a different one, never mind!

**Comment by tmb** [comment]
wrong PR, oops

File: app/views/invoices/index.html.erb [hunk 3]
```diff
@@ -43,8 +43,8 @@
             <span class="block h1 black line-height-1 <%= i.state %>"><%= render_money_short i.item_amount %></span>
           </div>
           <div class="flex flex-wrap justify-between items-center">
-              <span class="ml0 badge bg-<%= i.state %>">
-                <%= i.state_text %>
+              <span class="ml0 badge bg-<%= i.archived? ? 'pending' : i.state %>">
+                <%= i.archived? ? 'Archived' : i.state_text %>
```
Comment (lachlanjc) [inline app/views/invoices/index.html.erb:47]: This should absolutely be integrated into the `state_text` helper, not made as an exception in this view

File: app/assets/stylesheets/components/_cards.scss [hunk 4]
```diff
@@ -50,3 +50,18 @@
     margin-bottom: 0;
   }
 }
+
+.banner {
+  @extend .card;
+  max-width: 32rem;
+  padding: .5em;
```
Comment (lachlanjc) [inline app/assets/stylesheets/components/_cards.scss:57]: Using `em` is rarely the right choice for padding—switch it to `rem` here

File: app/assets/stylesheets/components/_cards.scss [hunk 5]
```diff
@@ -50,3 +50,18 @@
...
+}
+
+.banner--archived {
+  background-color: $yellow1;
+  border: 1px solid $yellow3;
+  color: $yellow8;
+  a {
+    color: $yellow8;
```
Comment (lachlanjc) [inline app/assets/stylesheets/components/_cards.scss:65]: Same here, can `inherit`

File: app/views/invoices/show.html.erb [hunk 6]
```diff
@@ -95,7 +105,7 @@
   <section class="card__banner border-top flex">
     <%= link_to 'View sent invoice', @invoice.hosted_invoice_url, class: 'btn bg-accent mr2', target: '_blank'  %>
     <%= link_to 'Download PDF', @invoice.invoice_pdf, class: 'btn bg-info', target: '_blank'  %>
-    <%= link_to "#{!@invoice.archived? ? 'Archive Invoice' : 'Un-archive Invoice'}", (!@invoice.archived? ? invoice_archive_path(@invoice) : invoice_unarchive_path(@invoice)), method: :post, class: 'btn bg-slate ml-auto' %>
+    <%= link_to "#{@invoice.archived? ? 'Un-archive' : 'Archive'}", (@invoice.archived? ? invoice_unarchive_path(@invoice) : invoice_archive_path(@invoice)), method: :post, class: 'btn bg-muted ml-auto' %>
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb:108]: Excellent change :+1:

File: app/assets/stylesheets/_dark.scss [hunk 7]
```diff
@@ -144,6 +144,11 @@ html[data-dark='true'] {
   .badge.bg-error {
     color: map-get($palette, error);
   }
+  .badge.archived {
+    background-color: rgba($yellow8, 0.625);
+    border: none;
```
Comment (lachlanjc) [inline app/assets/stylesheets/_dark.scss:149]: No badges have borders—can probably be removed

File: app/assets/stylesheets/_dark.scss [hunk 8]
```diff
@@ -144,6 +144,11 @@ html[data-dark='true'] {
   .badge.bg-error {
     color: map-get($palette, error);
   }
+  .badge.archived {
```
Comment (lachlanjc) [inline app/assets/stylesheets/_dark.scss:147]: Off the top of your head, what does the `.archived` class mean?

This pattern is inconsistent. On the `.banner` below, you use `.banner--archived`, & on most badges, we use `.bg-STATE`. Following either one of those patterns would make the purpose of this class clearer.

File: app/assets/stylesheets/_dark.scss [hunk 9]
```diff
@@ -183,4 +188,12 @@ html[data-dark='true'] {
   .shade-yellow {
     background-color: rgba($yellow4, 0.125);
   }
+  .banner--archived {
+    background-color: rgba($yellow1, 0.125);
+    border-color: rgba($yellow8, 0.625);
+    color: $yellow3;
+    a {
+      color: $yellow3;
```
Comment (lachlanjc) [inline app/assets/stylesheets/_dark.scss:196]: `inherit` would be a better choice here, so you're only declaring the color once

File: app/views/invoices/show.html.erb [hunk 10]
```diff
@@ -2,6 +2,16 @@
 <% page_md %>
 <%= render 'events/nav' %>
 
+<% if @invoice.archived? %>
+  <div class="banner banner--archived mx-auto mt3 flex items-center justify-center pl2 pr2 py1 line-height-1.5">
+      <div class="center">
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb:7]: Why not add the `center` class to the line above & remove this div?

Also, indentation is off here

File: app/views/invoices/show.html.erb [hunk 11]
```diff
@@ -2,6 +2,16 @@
 <% page_md %>
 <%= render 'events/nav' %>
 
+<% if @invoice.archived? %>
+  <div class="banner banner--archived mx-auto mt3 flex items-center justify-center pl2 pr2 py1 line-height-1.5">
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb:6]: `line-height-1.5` is not a class that does something :) search for the `line-height-1`, `line-height-4`, etc classes

File: app/controllers/invoices_controller.rb [hunk 12]
```diff
@@ -84,7 +84,6 @@ def archive
     @invoice.archive
 
     if @invoice.save
-      flash[:success] = 'Invoice has been archived.'
```
Comment (lachlanjc) [inline app/controllers/invoices_controller.rb]: :+1:

**Review by lachlanjc — COMMENTED** [review]
Great progress, made a few small notes on improving the frontend code.

File: app/views/invoices/show.html.erb [hunk 13]
```diff
@@ -5,7 +5,7 @@
 <% if @invoice.archived? %>
   <div class="banner banner--archived mx-auto mt3 flex items-center justify-center pl2 pr2 py1 line-height-1.5">
       <div class="center">
-        <span class="bold block">This invoice is archived.</span>
+        <span class="bold block">This invoice was archived by <%= @invoice.archived_by.name %> on <%= format_date @invoice.archived_at %></span>
```
Comment (lachlanjc) [inline app/views/invoices/show.html.erb:8]: Never refer only to a team member’s name, use the `user_mention` helper instead

File: app/models/invoice.rb [hunk 14]
```diff
@@ -108,14 +109,6 @@ def filter_data
     }
   end
 
-  def archive
```
Comment (lachlanjc) [inline app/models/invoice.rb]: I realize this code was deleted, but in the future, Ruby/Rails convention is to use a `!` at the end of any class that modifies the object, so you know `Invoice.archive!` changes the object directly

File: app/views/invoices/show.html.erb (hunk 13 above)
Comment (tmb) [reply app/views/invoices/show.html.erb:8]: user mention looks weird and disturbs the flow

**Pull request merged**